File: DataStructures/DataStructures/Stack.swift

class Stack<T> {
    private var buffer: DataStructureBuffer<T>

    init(capacity: Int = 1) {
        // TODO: Any kind of thread operations will very easily break
        buffer = DataStructureBuffer<T>(capacity: capacity)
    }
    
    func push(_ object: T) {
        buffer.increaseArrayCapacityIfNecessary()
        let pointer = buffer.pointer.advanced(by: buffer.count)
        buffer.add(object, to: pointer)
    }
    
    @discardableResult func pop() -> T {
        let indexToRemove = buffer.count - 1
        let pointer = buffer.pointer.advanced(by: indexToRemove)
        let object: T = pointer.pointee
        buffer.remove(pointer)
        return object
    }
    
    func count() -> Int {
        return buffer.count
    }
}

extension Stack where T: Equatable {
    func contains(_ object: T) -> Bool {
        for counter in 0..<buffer.count {
            if object == buffer.pointer.advanced(by: counter).pointee {
                return true
            }
        }
        return false
    }
    
}

File: CrackingCodeInterview/C03_03_StackOfPlates/src/Stack.java
import java.util.ArrayList;

public class Stack {
    private ArrayList<Node> array = new ArrayList<>();

    public Node pop() {
        if (array.size() == 0) {
            return null;
        }

        return array.remove(array.size() - 1);
    }

    public void push (Node node) {
        array.add(node);
    }

    public int size() {
        return array.size();
    }
}
